File: src/normalization/normalizer.py
"""
Bộ chuẩn hóa tiếng Việt chống né lọc (rule-based kết hợp từ điển teencode).

Xử lý 5 bước:
1. Chuẩn hóa Unicode NFC
2. Ánh xạ ký tự đồng dạng (lookalike)
3. Xóa ký tự phân cách xen kẽ chữ cái (.-_~·)
4. Rút gọn ký tự lặp kéo dài
5. Tra từ điển khôi phục teencode
"""

from __future__ import annotations

import json
import re
import unicodedata
from pathlib import Path

_DICT_DIR = Path(__file__).parent
_TEENCODE_DICT_PATH = _DICT_DIR / "teencode_dict.json"


def _load_teencode_dict() -> dict[str, str]:
    if not _TEENCODE_DICT_PATH.exists():
        return {}
    with open(_TEENCODE_DICT_PATH, "r", encoding="utf-8") as f:
        return json.load(f)


TEENCODE_MAP: dict[str, str] = _load_teencode_dict()

# Ánh xạ số/ký hiệu thay thế cho chữ cái tương tự
LOOKALIKE_MAP = {
    "4": "a", "3": "e", "1": "i", "0": "o",
    "5": "s", "9": "g", "8": "b", "@": "a",
}

_TOKEN_RE = re.compile(r"\S+")


def _map_lookalike_token(token: str) -> str:
    # Giữ nguyên số đo hoặc thông số kỹ thuật bắt đầu bằng số (ví dụ: 40km, 1TB)
    if token and token[0].isdigit():
        return token

    num_alpha = sum(1 for c in token if c.isalpha())
    num_convertible = sum(1 for c in token if c in LOOKALIKE_MAP)

    # Chỉ thay thế khi token chủ yếu là chữ cái
    if num_convertible == 0 or num_alpha <= num_convertible:
        return token

    return "".join(LOOKALIKE_MAP.get(ch, ch) for ch in token)


def map_lookalike_chars(text: str) -> str:
    """Thay thế ký tự đồng dạng trong từng token."""
    def replace_token(match):
        return _map_lookalike_token(match.group())
    return _TOKEN_RE.sub(replace_token, text)


# Chỉ xóa ký tự phân cách khi nó nằm giữa hai chữ cái tiếng Việt (ví dụ: n.g.u -> ngu, giữ nguyên 3.14)
SEPARATOR_PATTERN = re.compile(r"(?<=[a-zA-ZÀ-ỹ])[.\-_·~](?=[a-zA-ZÀ-ỹ])")


def remove_inserted_separators(text: str) -> str:
    """Xóa ký tự phân cách xen giữa các chữ cái."""
    return SEPARATOR_PATTERN.sub("", text)


ELONGATION_PATTERN = re.compile(r"([a-zA-ZÀ-ỹ])\1{2,}")


def _strip_base(char: str) -> str:
    """Lấy ký tự gốc không dấu (á -> a, ồ -> o)."""
    s = char.replace("đ", "d").replace("Đ", "D")
    nfkd = unicodedata.normalize("NFD", s)
    return "".join(c for c in nfkd if not unicodedata.combining(c))


def collapse_elongation(text: str) -> str:
    """
    Rút gọn ký tự lặp kéo dài (>= 3 lần) và bỏ ký tự gốc lặp phía sau ký tự có dấu.
    """
    text = ELONGATION_PATTERN.sub(r"\1", text)

    # Xóa ký tự gốc thừa sau ký tự có dấu (ví dụ: quáa -> quá)
    if len(text) >= 2:
        chars = list(text)
        result = [chars[0]]
        for i in range(1, len(chars)):
            prev = chars[i - 1]
            curr = chars[i]
            if (prev != curr
                    and curr.isalpha()
                    and prev.isalpha()
                    and _strip_base(prev) == _strip_base(curr)
                    and prev != _strip_base(prev)):
                continue
            result.append(curr)
        text = "".join(result)

    return text


_LEADING_PUNCT_RE = re.compile(r"^([^\w]*)")
_TRAILING_PUNCT_RE = re.compile(r"([^\w]*)$")


def _transfer_case(original: str, replacement: str) -> str:
    if not original or not replacement:
        return replacement
    if original[0].isupper() and replacement[0].islower():
        return replacement[0].upper() + replacement[1:]
    return replacement


def restore_teencode(text: str) -> str:
    """Tra từ điển và thay thế từ teencode về dạng chuẩn."""
    if not TEENCODE_MAP:
        return text

    words = text.split()
    result = []

    for w in words:
        lead = _LEADING_PUNCT_RE.match(w).group(1)
        trail = _TRAILING_PUNCT_RE.search(w).group(1)
        core = w[len(lead):len(w) - len(trail)] if trail else w[len(lead):]

        if not core:
            result.append(w)
            continue

        lookup = core.lower()
        if lookup in TEENCODE_MAP:
            replacement = _transfer_case(core, TEENCODE_MAP[lookup])
            result.append(lead + replacement + trail)
        else:
            result.append(w)

    return " ".join(result)


def normalize_unicode(text: str) -> str:
    return unicodedata.normalize("NFC", text)


def normalize(text: str) -> str:
    """
    Chạy tuần tự 5 bước chuẩn hóa:
    NFC -> Ký tự đồng dạng -> Ký tự phân cách -> Rút gọn lặp -> Từ điển teencode
    """
    text = normalize_unicode(text)
    text = map_lookalike_chars(text)
    text = remove_inserted_separators(text)
    text = collapse_elongation(text)
    text = restore_teencode(text)
    return text


def get_dict_size() -> int:
    return len(TEENCODE_MAP)


def reload_dict() -> None:
    global TEENCODE_MAP
    TEENCODE_MAP = _load_teencode_dict()


if __name__ == "__main__":
    import sys
    if hasattr(sys.stdout, "reconfigure"):
        try:
            sys.stdout.reconfigure(encoding="utf-8")
        except Exception:
            pass
    print(f"Từ điển: {get_dict_size()} mục\n")

    test_cases = [
        ("m4y n9u qu4", "Ký tự đồng dạng số sang chữ"),
        ("n.g.u q.u.á", "Ký tự phân cách xen kẽ"),
        ("m4.y n9.u", "Kết hợp đồng dạng và phân cách"),
        ("nàoooo đẹppppp", "Kéo dài ký tự"),
        ("quáaaaa", "Kéo dài sau ký tự có dấu"),
        ("ko bit j", "Viết tắt teencode"),
        ("cx đc thui", "Nhiều từ viết tắt"),
        ("ngu qua", "Thiếu dấu"),
        ("zậy fải k", "Viết theo phát âm"),
        ("400k", "Số và đơn vị giữ nguyên"),
        ("3.14", "Số thập phân giữ nguyên"),
        ("tôi thức dậy", "Câu bình thường giữ nguyên"),
        ("ko, tôi ko.", "Teencode kèm dấu câu"),
    ]

    for text, desc in test_cases:
        out = normalize(text)
        status = " (thay đổi)" if out != text else " (giữ nguyên)"
        print(f"  {desc:35s} | {text:25s} -> {out}{status}")

File: demo/js/normalizer.js
/**
 * Bộ chuẩn hóa tiếng Việt chống né lọc (chuyển thể từ normalizer.py sang JavaScript).
 *
 * Quy trình xử lý:
 * 1. Chuẩn hóa Unicode NFC
 * 2. Thay thế ký tự đồng dạng (số/ký hiệu thay cho chữ cái)
 * 3. Xóa ký tự phân cách xen giữa chữ cái (.-_~·)
 * 4. Rút gọn ký tự lặp kéo dài (>= 3 lần liên tiếp)
 * 5. Tra từ điển khôi phục teencode
 *
 * @module normalizer
 */

let TEENCODE_MAP = {};

/**
 * Tải từ điển teencode từ file JSON.
 * Cần gọi hàm này trước khi thực hiện chuẩn hóa có tra từ điển.
 */
export async function init(dictUrl = '../teencode_dict.json') {
  try {
    const resp = await fetch(dictUrl);
    if (resp.ok) {
      TEENCODE_MAP = await resp.json();
    } else {
      console.warn('[normalizer] Không tải được từ điển teencode, bỏ qua bước tra từ điển.');
    }
  } catch (e) {
    console.warn('[normalizer] Lỗi khi tải từ điển:', e.message);
  }
}

const VIET_LETTER = /[a-zA-Z\u00C0-\u024F\u1E00-\u1EFF]/;

function isVietnameseLetter(ch) {
  return VIET_LETTER.test(ch);
}

function normalizeUnicode(text) {
  return text.normalize('NFC');
}

// Bảng ánh xạ ký tự đồng dạng phổ biến
const LOOKALIKE_MAP = {
  '4': 'a', '3': 'e', '1': 'i', '0': 'o',
  '5': 's', '9': 'g', '8': 'b', '@': 'a',
};

function mapLookalikeToken(token) {
  // Giữ nguyên các token bắt đầu bằng số (ví dụ: 40km, 100k, 5G)
  if (token.length > 0 && /\d/.test(token[0])) {
    return token;
  }

  let numAlpha = 0;
  let numConvertible = 0;
  for (const ch of token) {
    if (isVietnameseLetter(ch)) numAlpha++;
    if (ch in LOOKALIKE_MAP) numConvertible++;
  }

  // Chỉ thay thế khi token chủ yếu là chữ cái
  if (numConvertible === 0 || numAlpha <= numConvertible) {
    return token;
  }

  return Array.from(token)
    .map(ch => (ch in LOOKALIKE_MAP ? LOOKALIKE_MAP[ch] : ch))
    .join('');
}

function mapLookalikeChars(text) {
  return text.replace(/\S+/g, match => mapLookalikeToken(match));
}

// Chỉ xóa ký tự phân cách khi nó nằm giữa hai chữ cái tiếng Việt (ví dụ: n.g.u -> ngu, còn 3.14 giữ nguyên)
const SEPARATOR_PATTERN = /(?<=[a-zA-Z\u00C0-\u024F\u1E00-\u1EFF])[.\-_·~](?=[a-zA-Z\u00C0-\u024F\u1E00-\u1EFF])/g;

function removeInsertedSeparators(text) {
  return text.replace(SEPARATOR_PATTERN, '');
}

// Rút gọn ký tự chữ cái lặp >= 3 lần (ví dụ: nàoooo -> nào)
const ELONGATION_PATTERN = /([a-zA-Z\u00C0-\u024F\u1E00-\u1EFF])\1{2,}/g;

function stripBase(ch) {
  let s = ch.replace(/đ/g, 'd').replace(/Đ/g, 'D');
  const nfkd = s.normalize('NFD');
  let result = '';
  for (const c of nfkd) {
    if (!/[\u0300-\u036F]/.test(c)) {
      result += c;
    }
  }
  return result;
}

function collapseElongation(text) {
  text = text.replace(ELONGATION_PATTERN, '$1');

  // Lược bỏ ký tự gốc lặp phía sau ký tự có dấu (ví dụ: quáa -> quá)
  if (text.length >= 2) {
    const chars = Array.from(text);
    const result = [chars[0]];
    for (let i = 1; i < chars.length; i++) {
      const prev = chars[i - 1];
      const curr = chars[i];
      if (
        prev !== curr &&
        isVietnameseLetter(curr) &&
        isVietnameseLetter(prev) &&
        stripBase(prev) === stripBase(curr) &&
        prev !== stripBase(prev)
      ) {
        continue;
      }
      result.push(curr);
    }
    text = result.join('');
  }

  return text;
}

const LEADING_PUNCT_RE = /^([^\w]*)/;
const TRAILING_PUNCT_RE = /([^\w]*)$/;

function transferCase(original, replacement) {
  if (!original || !replacement) return replacement;
  if (
    original[0] === original[0].toUpperCase() &&
    original[0] !== original[0].toLowerCase() &&
    replacement[0] === replacement[0].toLowerCase()
  ) {
    return replacement[0].toUpperCase() + replacement.slice(1);
  }
  return replacement;
}

function restoreTeencode(text) {
  if (Object.keys(TEENCODE_MAP).length === 0) return text;

  const words = text.split(' ');
  const result = [];

  for (const w of words) {
    const leadMatch = w.match(LEADING_PUNCT_RE);
    const trailMatch = w.match(TRAILING_PUNCT_RE);
    const lead = leadMatch ? leadMatch[1] : '';
    const trail = trailMatch ? trailMatch[1] : '';
    const core = trail ? w.slice(lead.length, w.length - trail.length) : w.slice(lead.length);

    if (!core) {
      result.push(w);
      continue;
    }

    const lookup = core.toLowerCase();
    if (lookup in TEENCODE_MAP) {
      const replacement = transferCase(core, TEENCODE_MAP[lookup]);
      result.push(lead + replacement + trail);
    } else {
      result.push(w);
    }
  }

  return result.join(' ');
}

/**
 * Thực hiện toàn bộ chuỗi chuẩn hóa văn bản.
 */
export function normalize(text) {
  text = normalizeUnicode(text);
  text = mapLookalikeChars(text);
  text = removeInsertedSeparators(text);
  text = collapseElongation(text);
  text = restoreTeencode(text);
  return text;
}
